rs/nns/governance/tla/Refresh_Neuron.tla

```diff
@@ -0,0 +1,164 @@
+    Neuron_Ids
+
+CONSTANTS
+    Refresh_Neuron_Process_Ids
+
+CONSTANTS
+    \* Minimum stake a neuron can have
+    MIN_STAKE,
+    \* The transfer fee charged by the ledger canister
+    TRANSACTION_FEE
+
+OP_ACCOUNT_BALANCE == "account_balance"
+ACCOUNT_BALANCE_FAIL == "Err"
+DUMMY_ACCOUNT == ""
+
+\* @type: (a -> b, Set(a)) => a -> b;
+Remove_Arguments(f, S) == [ x \in (DOMAIN f \ S) |-> f[x]]
+
+request(caller, request_args) == [caller |-> caller, method_and_args |-> request_args]
+account_balance(account) == Variant("AccountBalance", [account |-> account])
+
+
+(* --algorithm Governance_Ledger_Refresh_Neuron {
+
+variables
+
+    neuron \in [{} -> {}];
+    \* Used to decide whether we should refresh or claim a neuron
+    neuron_id_by_account \in [{} -> {}];
+    \* The set of currently locked neurons
+    locks = {};
+    \* The queue of messages sent from the governance canister to the ledger canister
+    governance_to_ledger = <<>>;
+    ledger_to_governance = {};
+    spawning_neurons = FALSE;
+
+macro refresh_neuron_reset_local_vars() {
+    account := DUMMY_ACCOUNT;
+    neuron_id := 0;
+}
+
+
+\* Modified from formal-models/tla/governance-ledger
+\* A Refresh_Neuron process simulates a call to refresh_neuron
+process ( Refresh_Neuron \in Refresh_Neuron_Process_Ids )
+    variable
+        \* There are two ways that the user can invoke a neuron refresh: 
+        \* 1. by specifying an account ID
+        \* 2. by specifying an existing neuron ID
+        \* We only model the second option; the second should follow from the invariant that 
+        \* \A nid aid : neuron_id_by_account[aid] = nid <=> neuron[nid].account = aid
+
+        \* The account is an argument; we let it be chosen non-deteministically
+        account = DUMMY_ACCOUNT;
+        \* The neuron_id is determined by account.
+        neuron_id = 0;
+    {
+    RefreshNeuron1:
+        either {
+            \* Simulate calls that just fail early and don't change the state.
+            \* Not so useful for model checking, but needed to follow the code traces.
+            goto Done;
+        } or {
+        with(nid \in DOMAIN(neuron) \ locks) {
+            neuron_id := nid;
+            account := neuron[nid].account;
+            locks := locks \union {neuron_id};
+            governance_to_ledger := Append(governance_to_ledger, request(self, account_balance(account)));
+        };
+        };
+    WaitForBalanceQuery:
+        \* Note that the "with" construct implicitly awaits until the set of values to draw from is non-empty
+        with(answer \in { resp \in ledger_to_governance : resp.caller = self }) {
+            ledger_to_governance := ledger_to_governance \ {answer};
+            if(answer.response /= Variant("Fail", UNIT)) {
+                with (b = VariantGetOrElse("BalanceQueryOk", answer.response, 0)) {
+                    if(b >= MIN_STAKE) {
+                        neuron := [neuron EXCEPT ![neuron_id] = [@ EXCEPT !.cached_stake = b] ]
```

Comment:
Apparently you can also write
```suggestion
                        neuron := [neuron EXCEPT ![neuron_id].cached_stake = b ]
```
(I wasn't aware of this syntax before).
Also, I'm thinking that we may want to have an invariant that the cached stake of a neuron subaccount doesn't go below min stake... Though I'm not sure if this always holds or we'd need some side conditions on it (for example, if the user deposits less than `MIN_STAKE` and tries to create a neuron, maybe it's temporarily violated)